### PR TITLE
Tests for scrolled-doc and scrolled-continuous

### LINKED ISSUE #159: testing scrolled-doc and scrolled-continuous for rs with scripting support
Statements in [§6.4](https://w3c.github.io/epub-specs/epub33/rs/#sec-scripted-content).

### PR DESCRIPTION
**Pull request by iherman**
The first is required, the second is optional for RS-s that support testing.

Fix #159

### CODE REVIEW

File: tests/scr-support_scrolled-doc/EPUB/content_001.xhtml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+			 */ 
+			function test() {
+			  const div_no_scripting = document.getElementById('no_scripting');
+			  div_no_scripting.style.display = 'none';
+			  const div_scripting = document.getElementById('scripting');
+			  div_scripting.style.display = 'block';
+			}
+			window.addEventListener('load', test);
+		</script>
+		<title>rendition:flow set to scrolled-doc when scripting is supported</title>
```
Comment (dlazin) [inline tests/scr-support_scrolled-doc/EPUB/content_001.xhtml:19]: This is a fine title but it's the *same* title as the other test, so please rename one of them.

IMO this is correct for the current test, but not properly describing the other test. I'm not actually sure which of the two SHOULDs in https://w3c.github.io/epub-specs/epub33/rs/#confreq-rs-scripted-scrolled the other test is meant to correspond to, but:

- if the first SHOULD, I think the other title should be "Reading system supports spine-level scripting with scrolled-doc"
- if the second SHOULD, I think the other title should be "rendition:flow set to scrolled-continuous when scripting is supported"

File: tests/scr-support_scrolled-doc/EPUB/content_001.xhtml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+	</head>
+	<body>
+		<section id="flow-015">
+			<h2>Scrolled Flow</h2>
+			<div id="no_scripting" style="display: block">
+				<p>This reading system does not support scripting. The test is not relevant, and can be marked as "null" in the report</p>
```
Comment (dlazin) [inline tests/scr-support_scrolled-doc/EPUB/content_001.xhtml:25]:
```suggestion
				<p>This reading system does not support scripting. The test is not relevant, and can be marked as "null" in the report.</p>
```

File: tests/scr-support_scrolled-continuous/EPUB/package.opf [hunk 3]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<package xmlns="http://www.idpf.org/2007/opf" version="3.0" unique-identifier="pub-id" prefix="rendition: http://www.idpf.org/vocab/rendition/#">
+  <metadata xmlns:dc="http://purl.org/dc/elements/1.1/">
+    <dc:coverage>Scripting</dc:coverage>
+    <dc:creator>Ivan Herman</dc:creator>
+    <dc:date>2022-06-02</dc:date>
+    <dc:description>Tests on handling overflow content with the "rendition:flow" property set to "scrolled-continuous". The publication is presented as one continuous scroll from spine item to spine item. This feature is optional if the reading system supports scripting.</dc:description>
```
Comment (dlazin) [inline tests/scr-support_scrolled-continuous/EPUB/package.opf:7]:
```suggestion
    <dc:description>Tests handling overflow content with the "rendition:flow" property set to "scrolled-continuous". The publication is presented as one continuous scroll from spine item to spine item. This feature is optional if the reading system supports scripting.</dc:description>
```

Comment (dlazin) [reply tests/scr-support_scrolled-continuous/EPUB/package.opf:7]: I'm not 100% sure about this description. See my question about *why* this requirement exists elsewhere — my understanding is that the spec is saying that if scripting is supported, the scrolling behavior actually changes? If so I think the description should say that, and if not I think we're still missing some other subtlety. (It doesn't make sense to me that the test would only be testing scrolling *if* the test requires scripting *and* the rendition:flow is explicitly set — I think the point is that support for scripting causes some change in behavior?)

Probably this is because I don't understand the difference between scrolled-doc and scrolled-continuous. The package says scrolled-doc. I think if scripting is supported, it's supposed to behave as if it were scrolled-doc? (That's what your title suggests.) If so I think the description should say that explicitly.

File: tests/scr-support_scrolled-doc/EPUB/content_001.xhtml [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+			</div>
+			<div id="scripting" style="display: none">
+				<p>
+					Tests whether setting the global <code>rendition:flow</code> property to <code>scrolled-continuous</code> results in
```
Comment (dlazin) [inline tests/scr-support_scrolled-doc/EPUB/content_001.xhtml:29]:
```suggestion
					Tests whether setting the global <code>rendition:flow</code> property to <code>scrolled-doc</code> results in
```

File: tests/scr-support_scrolled-doc/EPUB/content_002.xhtml [hunk 5]
```diff
@@ -0,0 +1,139 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE html>
+<html xmlns="http://www.w3.org/1999/xhtml" xmlns:epub="http://www.idpf.org/2007/ops" xml:lang="en" lang="en">
+	<head>
+		<meta charset="utf-8" />
+		<title>rendition:flow set to scrolled-doc when scripting is supported: chapter 1</title>
```
Comment (dlazin) [inline tests/scr-support_scrolled-doc/EPUB/content_002.xhtml:6]:
```suggestion
		<title>rendition:flow set to scrolled-doc when scripting is supported</title>
```

File: tests/scr-support_scrolled-doc/EPUB/nav.xhtml [hunk 6]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!DOCTYPE html>
+<html xmlns="http://www.w3.org/1999/xhtml" xmlns:epub="http://www.idpf.org/2007/ops" xml:lang="en" lang="en">
+	<head>
+		<meta charset="utf-8" />
+		<title>rendition:flow set to scrolled-doc when scripting is implemented</title>
```
Comment (dlazin) [inline tests/scr-support_scrolled-doc/EPUB/nav.xhtml:6]:
```suggestion
		<title>rendition:flow set to scrolled-doc when scripting is supported</title>
```

File: tests/scr-support_scrolled-doc/EPUB/content_003.xhtml [hunk 7]
```diff
@@ -0,0 +1,104 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE html>
+<html xmlns="http://www.w3.org/1999/xhtml" xmlns:epub="http://www.idpf.org/2007/ops" xml:lang="en" lang="en">
+	<head>
+		<meta charset="utf-8" />
+		<title>rendition:flow set to scrolled-doc when scripting is supported: chapter 2</title>
```
Comment (dlazin) [inline tests/scr-support_scrolled-doc/EPUB/content_003.xhtml:6]:
```suggestion
		<title>rendition:flow set to scrolled-doc when scripting is supported</title>
```

File: tests/scr-support_scrolled-continuous/EPUB/package.opf [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    <dc:identifier id="pub-id">scr-support_scrolled-continuous</dc:identifier>
+    <dc:language>en</dc:language>
+    <dc:publisher>W3C</dc:publisher>
+    <dc:title>rendition:flow set to scrolled-doc when scripting is supported</dc:title>
+    <link rel="dcterms:rights" href="https://www.w3.org/Consortium/Legal/2015/copyright-software-and-document"/>
+    <link rel="dcterms:rightsHolder" href="https://www.w3.org"/>
+    <meta property="belongs-to-collection">should</meta>
+    <meta property="dcterms:isReferencedBy">https://www.w3.org/TR/epub-rs-33/#confreq-rs-scripted-flow-scrolled</meta>
```
[the remark this answers is not in the document]
Comment (iherman) [reply tests/scr-support_scrolled-continuous/EPUB/package.opf:15]: Yes, this is https://github.com/w3c/epub-specs/pull/2318

File: tests/scr-support_scrolled-doc/EPUB/package.opf [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    <dc:creator>Ivan Herman</dc:creator>
+    <dc:date>2022-06-02</dc:date>
+    <dc:description>Tests on handling overflow content with the "rendition:flow" property set to "scrolled-doc". Each of the chapters in this publication should be vertically scrollable, and swiping horizontally should moves from one to the next. This feature is required if the reading system supports scripting.</dc:description>
+    <dc:identifier id="pub-id">scr-support_scrolled-doc</dc:identifier>
+    <dc:language>en</dc:language>
+    <dc:publisher>W3C</dc:publisher>
+    <dc:title>rendition:flow set to scrolled-doc when scripting is supported</dc:title>
+    <link rel="dcterms:rights" href="https://www.w3.org/Consortium/Legal/2015/copyright-software-and-document"/>
+    <link rel="dcterms:rightsHolder" href="https://www.w3.org"/>
+    <meta property="dcterms:isReferencedBy">https://www.w3.org/TR/epub-rs-33/#confreq-rs-scripted-flow-scrolled</meta>
```
[the remark this answers is not in the document]
Comment (iherman) [reply tests/scr-support_scrolled-doc/EPUB/package.opf:14]: As before, https://github.com/w3c/epub-specs/pull/2318

File: tests/scr-support_scrolled-continuous/EPUB/content_001.xhtml [hunk 10]
```diff
@@ -0,0 +1,44 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (iherman) [reply tests/scr-support_scrolled-continuous/EPUB/content_001.xhtml:1]: Not really. The statement that is being tested is:

> Similarly, if it supports spine-level scripting in reflowable EPUB content documents, it MUST implement the
"<code>scrolled-doc</code>" presentation mode and SHOULD implement the
"<code>scrolled-continuous</code>" presentation mode.

This is the span on which I have added an anchor in the separate PR. In fact, only one, namely scr-support_scrolled-doc is really necessary (this is the only MUST statement in the whole paragraph), I just added the other test to also test the SHOULD statement in the same sentence.

I could have separated the two statements, meaning cutting the same statements into two tests, but looked a bit silly...

In other words, these tests are not testing the full paragraph. I think the naming is fine as is...

---

I agree that the full spec sentence is a big convoluted. It says two different things:

1. *If* the RS supports scrolled-doc and scrolled-continuous _then_ it should support scripting
2. *if* the RD supports scripting _then_ it must support scrolled-doc and should support scrolled continuous

The only reason these tests exist is due to the presence of that lonely MUST statement. I must admit I do not know why that MUST is there and why it isn't only a SHOULD; maybe @dauwhe or @mattgarrish, as living histories of EPUB, could tell us...

File: tests/scr-support_scrolled-continuous/EPUB/content_001.xhtml [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+			 * 
+			 */ 
+			function test() {
+			  const div_no_scripting = document.getElementById('no_scripting');
+			  div_no_scripting.style.display = 'none';
+			  const div_scripting = document.getElementById('scripting');
+			  div_scripting.style.display = 'block';
+			}
+			window.addEventListener('load', test);
+		</script>
```
[the remark this answers is not in the document]
Comment (iherman) [reply tests/scr-support_scrolled-continuous/EPUB/content_001.xhtml:18]: It is not how the document _is_ presented but how the document _may be_ presented, depending on user setting. And I do not have an answer as for why these things are related. @dauwhe or @mattgarrish may have some hidden knowledge...

File: tests/scr-support_scrolled-continuous/EPUB/content_001.xhtml [hunk 12]
```diff
@@ -0,0 +1,44 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE html>
+<html xmlns="http://www.w3.org/1999/xhtml" xmlns:epub="http://www.idpf.org/2007/ops" xml:lang="en" lang="en">
+	<head>
```
[the remark this answers is not in the document]
Comment (iherman) [reply tests/scr-support_scrolled-continuous/EPUB/content_001.xhtml:4]: I copied this from another test, so I am not sure of the statement "all use 2-space indent". It may be true that all the tests I have made use tabs, actually... Let me look at this later for my tests in general...

(It reflects my editor setting, where tab is 4 spaces, and is all inherited from the days when I used to program in python...)

File: tests/scr-support_scrolled-doc/EPUB/content_001.xhtml (hunk 1 above)
Comment (iherman) [reply tests/scr-support_scrolled-doc/EPUB/content_001.xhtml:19]: Actually... the other tests' titles were wrong! It should have been scrolled-continuous and not scrolled-doc... (changed now)

File: tests/scr-support_scrolled-doc/EPUB/content_001.xhtml [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+					scrollable publication in the case when scripted is supported by the reading system. 
+					The test passes if the publication is presented as one continuous scroll from spine item to spine item.
```
[the remark this answers is not in the document]
Comment (iherman) [reply tests/scr-support_scrolled-doc/EPUB/content_001.xhtml:31]: Actually, I messed up the test descriptions (cut-and-paste effect. This is the description for the other test, and the description on the other one should be moved here. Done now...

File: tests/scr-support_scrolled-doc/EPUB/content_001.xhtml (hunk 4 above)
Comment (iherman) [reply tests/scr-support_scrolled-doc/EPUB/content_001.xhtml]: As I said, I messed up the two descriptions. Settled now.
